kge/job/train.py

```diff
@@ -530,108 +530,118 @@ def __init__(self, config, dataset, parent_job=None):
+        prev_end = 0
+        for query, use in self.queries.items():
+            if use:
+                data[query] = self.dataset.index(f"{self.train_split}_{query}")
+                self.query_end_index[query] = prev_end + len(data[query])
+                prev_end = self.query_end_index[query]
+
+                # convert indexes to pytorch tensors: a nx2 keys tensor (rows = keys),
+                # an offset vector (row = starting offset in values for corresponding
+                # key), a values vector (entries correspond to values of original
+                # index)
+                #
+                # Afterwards, it holds:
+                # index[keys[i]] = values[offsets[i]:offsets[i+1]]
 
+                (
+                    self.keys[query],
+                    self.values[query],
+                    self.offsets[query]
+                ) = prepare_index(data[query])
+
+        self.num_examples = prev_end
         # create dataloader
         self.loader = torch.utils.data.DataLoader(
-            range(len(train_sp) + len(train_po)),
+            range(self.num_examples),
             collate_fn=self._get_collate_fun(),
             shuffle=True,
             batch_size=self.batch_size,
             num_workers=self.config.get("train.num_workers"),
             pin_memory=self.config.get("train.pin_memory"),
         )
-        self.num_examples = len(train_sp) + len(train_po)
 
     def _get_collate_fun(self):
-        num_sp = len(self.train_sp_keys)
-
         # create the collate function
         def collate(batch):
-            """For a batch of size n, returns a triple of:
+            """For a batch of size n, returns a dictionary of:
 
-            - pairs (nx2 tensor, row = sp or po indexes),
-            - label coordinates (position of ones in a batch_size x num_entities tensor)
-            - is_sp (vector of size n, 1 if corresponding example_index is sp, 0 if po)
+            - pairs (nx2 tensor, row = sp, po or so indexes),
+            - label coordinates (position of ones in a n x num_entities or n x num_relations tensor for so respectively)
+            - query_type (vector of size n, 2 if corresponding example_index is so, 1 if sp, 0 if po)
             - triples (all true triples in the batch: needed for weighted penalties only)
 
             """
+
             # count how many labels we have
             num_ones = 0
             for example_index in batch:
-                if example_index < num_sp:
-                    num_ones += self.train_sp_offsets[example_index + 1]
-                    num_ones -= self.train_sp_offsets[example_index]
-                else:
-                    example_index -= num_sp
-                    num_ones += self.train_po_offsets[example_index + 1]
-                    num_ones -= self.train_po_offsets[example_index]
+                prev_end = 0
+                for query, end in self.query_end_index.items():
+                    if example_index < end:
```

Comment:
relies on dict to be in some order (which is always the case in python)

Reply:
Is it? While I do think that this is the case in any likely implementation as long as the keyset is not changed, making such an assumption makes the code hard to read. The value of "end" depends on the order of the keys in the iteration.

More readable may be to use a map from query to (begin,end) or an ordered array of (end_index, query_type) pairs. An ordereddict would also be fine, but that still makes the code somewhat unaccessible.

Reply:
I had another look. The code sets `self.query_end_index[query] = prev_end + len(data[query])`. It implicitly assumes that `query_end_index` retains insertion order / is ordered the same way as `queries`. That's a hard-to-grasp assumption and it may not be valid.

Reply:
As far as I can see insertion into `self.query_end_index` is based on a loop over `queries` and `self.query_end_index` is a dict, therefore retaining insertion order according to the order in queries. Anyhow, I will adjust the code to be easier to understand